site/_posts/2020-05-20-building-protected-api-with-rails-and-jwt.md

```diff
@@ -0,0 +1,254 @@
+Rails.application.routes.draw do
+  resources :messages, only: [:index]
+end
+```
+
+The `Messages` controller won't be too complicated. It returns a hardcoded list of messages when a `GET` request is made to the `/messages` path. In a real world application, of course, you would store messages in the database and pull them dynamically using ActiveRecord. But for this tutorial, a hardcoded list suffices.
+
+Create the controller at `app/controllers/messages_controller.rb`. Here is what the class looks like:
+
+```ruby
+class MessagesController < ApplicationController
+  def index
+    messages = []
+    messages << "Hello"
+    render json: { messages: messages }.to_json, status: :ok
+  end
+end
+```
+
+If you start up your Rails server:
+
+```shell
+rails s -p 4000
+```
+
+You should now be able to visit `http://localhost:4000/messages` and see some messages:
+
+```json
+{"messages":["Hello"]}
+```
+
+But let's add a test so future changes don't cause any surprises. Create the controller test at `test/controllers/messages_controller_test.rb`. Here are the contents of that class:
+
+```ruby
+require 'test_helper'
+
+class MessagesTest < ActionDispatch::IntegrationTest
+  test "can get messages" do
+    get "/messages"
+    assert_response :success
+  end
+  test "can get messages content" do
+    get "/messages"
+    res = JSON.parse(@response.body)
+    assert_equal '{"messages"=>["Hello"]}', res.to_s
+  end
+end
+```
+
+Now we can run our test and make sure that we are getting what we expect:
+
+```shell
+$ rails test test/integration/messages_test.rb
+Running via Spring preloader in process 15492
+Run options: --seed 1452
+
+# Running:
+
+..
+
+Finished in 0.119373s, 16.7542 runs/s, 16.7542 assertions/s.
+2 runs, 2 assertions, 0 failures, 0 errors, 0 skips
+```
+
+Excellent! We have a working API which returns well formed JSON! Rails even takes care of setting the `Content-Type` header to `application/json; charset=utf-8`. Now let's secure the API.
+
+## Secure the API
+
+As a reminder, we're going to use a JWT to secure this API. While you can secure Rails APIs using [a variety of methods](https://edgeguides.rubyonrails.org/action_controller_overview.html#http-authentications), using a token has advantages. You can integrate with a number of identity providers that offer OAuth or SAML support, which allows you to leverage an existing robust identity management system to control API access. You can also embed additional metadata into a JWT, including attributes like roles.
+
+The first step to changing this API is to write tests. Let's modify the test to provide a JWT and expect `:forbidden` HTTP statuses when the token doesn't meet our expectations.
+
+```ruby
+class MessagesTest < ActionDispatch::IntegrationTest
+  test "can' get messages with no auth" do
+    get "/messages"
+    assert_response :forbidden
+  end
+  test "can get messages with header" do
+    get "/messages", headers: { "HTTP_AUTHORIZATION" => "Bearer " + build_jwt }
```

Comment:
Is "HTTP_AUTHORIZATION" a Rails thing that translates to the standard Authorization header? We should probably use the standard header if not.

Reply:
It does translate to the authorization header. https://www.rubydoc.info/docs/rails/6.0.2.1/ActionDispatch/Request#authorization-instance_method

Looks like magic rails has done for a while: https://stackoverflow.com/questions/2311883/authorization-header-in-ruby-on-rails-accessed-with-key-http-authorization-inste